foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/mappings/OneToManyMapping.java

```diff
@@ -982,19 +1045,56 @@ public void postInsert(WriteObjectQuery query) throws DatabaseException, Optimis
-                    updateTargetRowPostInsertSource(query);
+                if ((requiresDataModificationEvents() || containerPolicy.shouldUpdateForeignKeysPostInsert())) {
+                    if (shouldDeferInsert()) {
+                        // Hey I might actually want to use an inner class here... ok array for now.
+                        Object[] event = new Object[2];
+                        event[0] = PostInsert;
+                        event[1] = query;
+                        query.getSession().getCommitManager().addDataModificationEvent(this, event);
+                    } else {
+                        ContainerPolicy cp = getContainerPolicy();
+                        Object objects = getRealCollectionAttributeValueFromObject(query.getObject(), query.getSession());
+                        if (cp.isEmpty(objects)) {
+                            return;
+                        }
+
+                        prepareTranslationRow(query.getTranslationRow(), query.getObject(), query.getDescriptor(), query.getSession());
+
+                        AbstractRecord keyRow = buildKeyRowForTargetUpdate(query);
+
+                        // Extract target field and its value. Construct insert
+                        // statement and execute it
+                        int size = targetPrimaryKeyFields.size();
+                        ClassDescriptor referenceDesc = getReferenceDescriptor();
+                        AbstractSession session = query.getSession();
+                        int objectIndex = 0;
+                        for (Object iter = cp.iteratorFor(objects); cp.hasNext(iter);) {
+                            AbstractRecord row = new DatabaseRecord();
+                            row.mergeFrom(keyRow);
+                            Object wrappedObject = cp.nextEntry(iter, query.getSession());
+                            Object object = cp.unwrapIteratorResult(wrappedObject);
+                            AbstractRecord databaseRow = referenceDesc.getObjectBuilder().buildRow(row, object, session, WriteType.INSERT);
+                            ContainerPolicy.copyMapDataToRow(cp.getKeyMappingDataForWriteQuery(wrappedObject, session), databaseRow);
+                            if (listOrderField != null) {
+                                databaseRow.put(listOrderField, objectIndex++);
+                            }
+                            InsertObjectQuery insertQuery = getInsertObjectQuery(session, referenceDesc);
+                            insertQuery.setObject(object);
+                            insertQuery.setCascadePolicy(query.getCascadePolicy());
+                            insertQuery.setTranslationRow(databaseRow);
+                            insertQuery.setModifyRow(databaseRow);
+                            insertQuery.setIsPrepared(false);
+                            query.getSession().executeQuery(insertQuery);
```

Comment:
Interesting, I tried something similar to this, but I struggled with was this portion of logic: https://github.com/eclipse-ee4j/eclipselink/blob/c63079c2e0b2c9bb568698abc313bb5f84f5acde/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/sessions/CommitManager.java#L139-L142

Here, we are looping through the classes, in order (Post, Comment, ect), and execute any inserts for new objects. This means, we first INSERT POST, then create any data modification events (postInsert events for instance). Then we INSERT COMMENT objects in the second iteration. I kept hitting the issue that this sort of change would end up calling INSERTs for COMMENT twice... However, I just tried this change and I am not getting that issue like I had been -_-' 

I'm wondering if we can just update 'updateTargetRowPostInsertSource()' to work like this and then change the logic so that 'updateTargetRowPostInsertSource()' can just be called. I think the main difference might be `insertQuery.setObject(object);`? That removes the objects from the uowChangeSet??

Reply:
Or better - the object is marked as committed during the first insert, so the commit manager knows it should not commit it again

Reply:
I do like this implementation better than mine. I am just looking it over to see if there is anything else to note, but I think we should go with this.